### PR TITLE
Speed up list page

### PR DESCRIPTION
**Pull request by michield**
## Description
With very large subscriber numbers, the list page is slow to load. Load the subscriber numbers asynchronously.

### CODE REVIEW

**Comment by michield** [comment]
The list page is very slow to load when there are millions of subscribers. This makes the loading of the counters in the page asynchronous, for a slightly better UX. A better solution would be to put the counters in the DB and have a cron calculate them every so often, but that's more work.

File: public_html/lists/admin/actions/listmembercount.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+$status = ' ';
+$cacheTimeout = rand(900,15000); ## randomly timeout the cache
+$now = time();
+
+if (isset($_SESSION['listmembercount'][$listid]['content']) && (($now - $_SESSION['listmembercount'][$listid]['lastupdate']) < $cacheTimeout)) {
+  $status = '<!-- cached -->'.$_SESSION['listmembercount'][$listid]['content'];
+} else {
+  sleep($_SESSION['listcounter'] * 30);
+  $_SESSION['listcounter']++;
```
Comment (bramley) [inline public_html/lists/admin/actions/listmembercount.php:63]: Might be a problem with several requests each trying to read and update a shared session variable.

**Comment by bramley** [comment]
The javascript and php delays need to be appropriate both to those with large number of subscribers, and those for whom there is not currently a problem but I'm not sure how that might be done. Maybe have a config value that would default to 0.

A different approach is to avoid trying to estimate how long each list will take to query by having a queue of requests. When each request is completed the next one is started, so they are made sequentially and do not overlap which I assume is the purpose of the delays.

A quick search showed this example of sequential loading that might be useful

https://stackoverflow.com/questions/14653202/jquery-load-content-sequentially
There is a jquery example further down the page
https://stackoverflow.com/a/14653313

**Comment by bramley** [comment]
Based on the SO topic, this js loads the divs sequentially. I put a delay to have an observable gap between each request

    if (typeof asyncLoadDiv != 'undefined') {
        var xindex = 0;
        triggerLoad();
    }
    function triggerLoad() {
        var url
        var div;

        if (xindex < asyncLoadDiv.length) {
            div = asyncLoadDiv[xindex];
            url = asyncLoadUrl[xindex];
            $("#"+div).load(url, triggerLoad);
            ++xindex;
            //~ var t = new Date().getTime();
            //~ while (new Date().getTime() < t + 1000);
        }
    }

The sleep in the php code was also removed.

**Comment by bramley** [comment]
This is a bit neater

```
function loadDivContent(index) {
    var div;
    var url;

    if (index < asyncLoadDiv.length) {
        div = asyncLoadDiv[index];
        url = asyncLoadUrl[index];
        $("#"+div).html(busyImage + '<span class="loadingprogressbanner"></span>');
        $("#"+div).load(url, function() {
            loadDivContent(index + 1);
        });
    }
}
```
and the process is initiated by

    if (typeof asyncLoadDiv != 'undefined') {
        loadDivContent(0);
    }

**Comment by michield** [comment]
Yes, that's nice. I'll give it a try.

**Comment by michield** [comment]
I have updated it with those suggestions, I think that works very nicely and it's also a method that will work for both big and small installation. There may be a few more places that can do with this approach.

File: public_html/lists/admin/actions/listmembercount.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+if (!defined('PHPLISTINIT')) {
+    die();
+}
+verifyCsrfGetToken();
+
+$status = ' ';
+
+if (empty($_SESSION['adminloggedin'])) return;
+
+$now = time();
```
Comment (bramley) [inline public_html/lists/admin/actions/listmembercount.php:14]: This assignment also is duplicated lower down.

Comment (michield) [reply public_html/lists/admin/actions/listmembercount.php]: Have removed the duplicates

**Comment by michield** [comment]
I think this works ok now, so marking as ready

**Review by aulona1 — APPROVED** [review]

**Pull request merged**